src/Core/Field.h

```diff
@@ -102,6 +102,37 @@ struct AggregateFunctionStateData
     }
 };
 
+struct CustomType
+{
+    struct CustomTypeImpl
+    {
+        virtual ~CustomTypeImpl() = default;
+        virtual const char * getTypeName() const = 0;
+        virtual String toString() const = 0;
+
+        virtual bool operator < (const CustomTypeImpl &) const = 0;
+        virtual bool operator <= (const CustomTypeImpl &) const = 0;
+        virtual bool operator > (const CustomTypeImpl &) const = 0;
+        virtual bool operator >= (const CustomTypeImpl &) const = 0;
+        virtual bool operator == (const CustomTypeImpl &) const = 0;
+    };
+
+    CustomType() = default;
+    explicit CustomType(std::shared_ptr<CustomTypeImpl> impl_) : impl(impl_) {}
```

Comment:
It should be rather `std::shared_ptr<const CustomTypeImpl>` as you copy that `impl`.